src/daft-io/src/object_io.rs

```diff
@@ -112,10 +117,203 @@ pub(crate) trait ObjectSource: Sync + Send {
+                                    if full_glob_matcher.is_match(fm.filepath.as_str())
+                                        && matches!(fm.filetype, FileType::File)
+                                    {
+                                        result_tx.send(Ok(fm)).await.expect("Internal multithreading channel is broken: results may be incorrect");
+                                    }
+                                }
+                                Err(super::Error::NotFound { .. }) => {}
+                                Err(e) => {
+                                    result_tx.send(Err(e)).await.expect("Internal multithreading channel is broken: results may be incorrect");
+                                }
+                            }
+                        }
+                    }
+                    Err(e) => result_tx.send(Err(e)).await.expect(
+                        "Internal multithreading channel is broken: results may be incorrect",
+                    ),
+                };
+
+            // BASE CASE: current fragment is the last fragment in `glob_fragments`
+            } else if i == glob_fragments.len() - 1 {
+                // Last fragment contains a wildcard: we list the last level and match against the full glob
+                if current_fragment.has_special_character() {
+                    let next_level_file_metadata =
+                        source.iter_dir(path.as_str(), Some("/"), None).await;
+
+                    match next_level_file_metadata {
+                        Ok(mut next_level_file_metadata) => {
+                            while let Some(fm) = next_level_file_metadata.next().await {
+                                match fm {
+                                    Ok(fm) => {
+                                        if matches!(fm.filetype, FileType::File)
+                                            && full_glob_matcher.is_match(fm.filepath.as_str())
+                                        {
+                                            result_tx.send(Ok(fm)).await.expect("Internal multithreading channel is broken: results may be incorrect");
+                                        }
+                                    }
+                                    Err(super::Error::NotFound { .. }) => (),
+                                    Err(e) => {
+                                        result_tx.send(Err(e)).await.expect("Internal multithreading channel is broken: results may be incorrect");
+                                    }
+                                }
+                            }
+                        }
+                        Err(e) => result_tx.send(Err(e)).await.expect(
+                            "Internal multithreading channel is broken: results may be incorrect",
+                        ),
+                    }
+                // Last fragment does not contain wildcard: we return it if the full path exists and is a FileType::File
```

Comment:
why do we assume it's a single file here? couldn't it also be a directory?
When I use s3fs i get:
<img width="582" alt="image" src="https://github.com/Eventual-Inc/Daft/assets/2550285/074e14bd-e37f-49d3-9190-d091c3b47dad">

Reply:
vs this branch
<img width="459" alt="image" src="https://github.com/Eventual-Inc/Daft/assets/2550285/ca9c2275-1544-4f61-9833-3a0650dfff62">

Reply:
I decided to not follow fsspec behavior, which simplifies the glob logic quite a bit and is more similar to a Unix glob

Our glob:
1. Only returns `File` matches (this lets us fall back onto prefix list without worrying about having to return intermediate dirs)
2. Doesn't match "folders" (this is more consistent with Unix glob behavior):

Using Python glob on a local machine:
```
import glob

# Our glob doesn't return "directories", so the equivalent behavior would be to return []
glob.glob("src/")
>>> ['src/']

glob.glob("src/*")
>>> ["src/daft-table", ...]
```

Reply:
Fixed by falling back on normal `ls` behavior for globs on paths with no special characters